### DISCUSSION TITLE
Multi argument typeof/nameof

### DISCUSSION THREAD

**Discussion by wrexbe** [dotnet/csharplang#7506] [Language Ideas]
Allow typeof, and nameof to take multiple arguments.
// Before
var types = new []{typeof(string), typeof(int), typeof(bool)};

// After
var types = typeof(string, int, bool);

Places that this might be used

- Convert LoadComponentsAsync into for loops
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/VisualStudio/Core/Def/RoslynPackage.cs#L212C12-L212C12
- Pass an array to the params of AddParts
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/Workspaces/CoreTest/TestCompositionTests.cs#L18
- Ignoring performance, all the null ServiceDescriptors could loop, and add to a dictionary
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/Workspaces/Remote/Core/ServiceDescriptors.cs#L55
- Init Hashset from the typeof array
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/VisualStudio/Core/Impl/SolutionExplorer/BaseItem.cs#L73
- Move repeated typeofs into a separate call to AddAssemblies
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/EditorFeatures/TestUtilities/EditorTestCompositions.cs#L30

Different code base:
- Use typeof, and ToList it.
https://github.com/space-wizards/space-station-14/blob/646869afa7ae9a1a1611458616fd8d4f435ce181/Content.Shared/Verbs/Verb.cs#L217
- Use the new typeof directly
https://github.com/space-wizards/space-station-14/blob/646869afa7ae9a1a1611458616fd8d4f435ce181/Content.Client/Paper/UI/PaperWindow.xaml.cs#L36
- Use the new typeof directly
https://github.com/liuxuaner/Unity3DTraining/blob/0a701a655c5c6e6d61ce246a76ad95cbeb1ce3fa/UGUITraining/UIViewBindDemo/Editor/UIExporter.cs#L14
- Use the new typeof directly
https://github.com/mrvux/MonoGame-Samples/blob/354804365d5d36fb76fcf45698218c0f5b5d5d24/StarWarrior/Game1.cs#L74
- Use the new typeof directly
https://github.com/LokiMidgard/commandline/blob/12aac090396f9380c97f82b35ab139b81070878d/src/CommandLine/Core/ReflectionExtensions.cs#L189C31-L189C31
- Use the new typeof directly
https://github.com/youhide/ServUO/blob/ae978cbdba8dea70281e356f5800626c1be06b75/Scripts/Mobiles/Named/Miasma.cs#L70

**Comment by HaloFour** [comment]
IMO, this feels like something that is very uncommon for a developer to need or want to do, to the extent that it doesn't feel worthwhile supporting as a language feature.

**Reply by wrexbe** [reply]
Alright, ignoring the lack of proof either way, is this a more doable solution?

**Reply by wrexbe** [reply]
Saying something is uncommon doesn't really make sense. I can't prove something that doesn't exist will get used a lot. I also can't prove it based on patterns people are currently using, because they may just be avoiding a pattern, because it's currently a pain.

**Reply by HaloFour** [reply]
I'm saying that the use case of needing to obtain an array of `System.Type` from literals known at compile-time is uncommon.  I'm also saying that the existing syntax isn't remotely onerous enough to prevent people from following that pattern today.

Language features are _enormously_ expensive to implement.

**Reply by wrexbe** [reply]
Maybe features are only enormously expensive, because C# keeps picking big headline grabbing features. Instead of picking the low hanging fruit. The big things grab attention, but the little problems that are already solved, when you run into them are what keep you around.

**Reply by CyrusNajmabadi** [reply]
What features would you cut to add this feature in?

**Reply by HaloFour** [reply]
The process itself is very expensive, requiring many thousands of hours to go through the design phase, Implementation, testing, documentation and eternal support.

https://learn.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points

**Reply by wrexbe** [reply]
> What features would you cut to add this feature in?

Collection expressions

**Reply by wrexbe** [reply]
> The process itself is very expensive, requiring many thousands of hours to go through the design phase, Implementation, testing, documentation and eternal support.

Maybe the problem isn't the features then, but the process. Maybe improving the process should be on the roadmap.

**Reply by CyrusNajmabadi** [reply]
> Collection expressions

Collection expressions have much more value and applicability than this proposal. Virtually all programs benefit from the former, while only a tiny tiny fraction would benefit from this, and likely only in a tiny handful of places. 

The bang/buck radio is dominating here between the two.

**Reply by CyrusNajmabadi** [reply]
We also base decisions on the reception we get from the community and ecosystem as a whole. Collection expressions have received a ton of virtually all positive sentiment. With broad feedback showing support. It won't be a very bitter pill for the ecosystem to cut it for this feature.

**Reply by HaloFour** [reply]
The process is intentionally conservative.  C# is a 22 year old language with millions of developers and trillions of lines of code.  The language team is very careful to ensure that it evolves in a sustainable way so that it remains usable and useful for another 22 years.  This is true of pretty much all languages.

**Comment by wrexbe** [comment]
I didn't take very long to find one place it could be used
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/docs/wiki/Getting-Started-C%23-Syntax-Transformation.md?plain=1#L485

// New code
```
List<MetadataReference> refs = new List<MetadataReference>();
foreach(var type in typeof(object, SyntaxTree, CSharpSyntaxTree))
{
    refs.Add(MetadataReference.CreateFromFile(type.Assembly.Location));
}
```

And here are more:
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/VisualStudio/Core/Def/RoslynPackage.cs#L212C12-L212C12
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/Workspaces/CoreTest/TestCompositionTests.cs#L18
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/Workspaces/Remote/Core/ServiceDescriptors.cs#L55
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/VisualStudio/Core/Impl/SolutionExplorer/BaseItem.cs#L73
https://github.com/dotnet/roslyn/blob/1cabbfeac0719ba84689348cc8db2a7a167ae14b/src/EditorFeatures/TestUtilities/EditorTestCompositions.cs#L30

Different code base:
https://github.com/space-wizards/space-station-14/blob/646869afa7ae9a1a1611458616fd8d4f435ce181/Content.Shared/Verbs/Verb.cs#L217
https://github.com/space-wizards/space-station-14/blob/646869afa7ae9a1a1611458616fd8d4f435ce181/Content.Client/Paper/UI/PaperWindow.xaml.cs#L36
https://github.com/liuxuaner/Unity3DTraining/blob/0a701a655c5c6e6d61ce246a76ad95cbeb1ce3fa/UGUITraining/UIViewBindDemo/Editor/UIExporter.cs#L14
https://github.com/mrvux/MonoGame-Samples/blob/354804365d5d36fb76fcf45698218c0f5b5d5d24/StarWarrior/Game1.cs#L74
https://github.com/LokiMidgard/commandline/blob/12aac090396f9380c97f82b35ab139b81070878d/src/CommandLine/Core/ReflectionExtensions.cs#L189C31-L189C31
https://github.com/youhide/ServUO/blob/ae978cbdba8dea70281e356f5800626c1be06b75/Scripts/Mobiles/Named/Miasma.cs#L70

**Reply by HaloFour** [reply]
Sure, I'm not saying that it's not useful _anywhere_, but is it useful in enough places to justify the cost of a language feature.  I severely doubt it.

**Reply by HaloFour** [reply]
Also, from looking at those examples, it seems like none of them would benefit from this syntax.  None of them are trying to make a `System.Type[]`.  From the few that could be potentially refactored I don't see a reason to want to.  There's very little code savings, if any.

**Reply by wrexbe** [reply]
It would be nice for bulk injection registrations, and in component systems for finding an entity with a set of components. The nameof could be useful in cases where there is a combination of arguments that cause the issue.

**Comment by En3Tho** [comment]
This could be easily implemented as a library:
```csharp
public Type[] TypesOf<T1, T2, T3>() => new[] { typeof(T1), typeof(T2), typeof,(T3) }
... Other overloads
```

I'm not sure why would you want a language feature for this kind of thing.

You can create a library like that but I doubt it will be used much because the code like that is written in 0.00000001% of time. It's not worth optimizing dev time on this kind of thing.

Collection literals on the other hand can easily save dev time, make collection initialization more succinct. Just like pattern matching is much more capable and expressive than bunch of nested if else blocks.
